docs/dev-guide/guide.md

```diff
@@ -567,6 +567,15 @@ To automatically fix lint and formatting errors, run `make format`.
 
 ![Question lifecycle flowchart](./question-flow.d2)
 
+## Assertions
+
+Depending on the context, we use different types of assertions.
+
+- In tests, we use the exported helpers from `vitest`, e.g. `assert.ok` or `assert.isDefined`.
+- In server code, to enforce invariants (e.g. something that should never happen), we use `assert` from `node:assert`.
+- If this is not an invariant, you can use `assertDefined`, or throw your own error.
```

Comment:
I don't really understand the difference. When would `assertDefined` _not_ be used to assert an invariant?